### PR TITLE
Several updates to the Gitlab CI

### PR DESCRIPTION
**Pull request by benson31**
- Automatically check for Al/H updates and trigger rebuilds
- Fix issues with coverage reporting
- Allow users to specify compilers more explicitly
- Source code flow and style improvements
- More useful output to stdout

### CODE REVIEW

File: .gitlab/build-and-test.sh [hunk 1]
```diff
@@ -64,27 +68,59 @@ echo "~~~~~          Host: ${hostname}"
...
+    h_head=$(fetch-sha elemental hydrogen)
+    h_latest="<not found>"
+    if [[ -f "${prefix}/h-latest.txt" ]]
+    then
+        h_latest=$(cat ${prefix}/h-latest.txt)
+    fi
+
+    if [[ "${al_head}" != "${al_latest}" ]]
+    then
+        echo "Aluminum hashes do not match!"
```
Comment (ndryden) [inline .gitlab/build-and-test.sh]: In a few months, I'm probably going to forget what this indicates and worry something went wrong. Maybe "Prebuilt Aluminum hash does not match latest, rebuilding" or something.

File: .gitlab/build-and-test.sh [hunk 2]
```diff
@@ -64,27 +68,59 @@ echo "~~~~~          Host: ${hostname}"
...
+        echo "  (head: ${al_head}; latest: ${al_latest})"
+        rebuild_deps=1
+    fi
+    if [[ "${h_head}" != "${h_latest}" ]]
+    then
+        echo "Hydrogen hashes do not match!"
```
Comment (ndryden) [inline .gitlab/build-and-test.sh]: Ditto.

**Review by ndryden — APPROVED** [review]
:shipit:

File: .gitlab/build-and-test.sh (hunk 1 above)
Comment (benson31) [reply .gitlab/build-and-test.sh]: Updated.

File: .gitlab/build-and-test.sh (hunk 2 above)
Comment (benson31) [reply .gitlab/build-and-test.sh]: Ditto.

**Pull request merged**